### PR TITLE
Fix crash on parquet column type mismatch

### LINKED ISSUE #68403: Parquet query failed with `In` decimal column filter
After https://github.com/ClickHouse/ClickHouse/pull/68135,  let's execute the following sql:

```sql 
CREATE TABLE file_engine_table (a DECIMAL(9,2)) ENGINE=File(Parquet);
INSERT INTO file_engine_table VALUES (2.00), (4.01), (5.00) , (10.00);
select * from file_engine_table where a in (5.00, 6.00, 8.00, 10.000, 20.00);
```

we get following error message:
``` shell
Code: 170. DB::Exception: Bad get: has Decimal256, requested Decimal32: (in file/uri ./store/769/76981421-d626-4639-b709-fd14c149a5ad//data.Parquet): While executing ParquetBlockInputFormat: While executing File. (BAD_GET) (version 24.9.1.1)
```

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed crash in Parquet filtering when data types in the file substantially differ from requested types (e.g. `... FROM file('a.parquet', Parquet, 'x String')`, but the file has `x Int64`). Without this fix, use `input_format_parquet_filter_push_down = 0` as a workaround.

ParquetBlockInputFormat didn't cast Field-s to the requested types because I thought KeyCondition allows any types. I don't remember why I thought that. It was crashing in `MergeTreeSetIndex::checkInRange()` when `FieldValue::update()` tried to insert Field of wrong type into a column.

Fixes #68403

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 61befa33a40d4ec78ba124107e700d14f9be9599 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68131/61befa33a40d4ec78ba124107e700d14f9be9599/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/21f8b03789220648948a0d545989d77d9183a927/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by jkartseva — APPROVED** [review]
Looks good

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 1]
```diff
@@ -758,6 +758,15 @@ static ColumnWithTypeAndName readNonNullableColumnFromArrowColumn(
                     case TypeIndex::IPv6:
                         return readIPv6ColumnFromBinaryData(arrow_column, column_name);
                     /// ORC format outputs big integers as binary column, because there is no fixed binary in ORC.
+                    ///
+                    /// When ORC/Parquet file says the type is "byte array" or "fixed len byte array",
+                    /// but the clickhouse query says to interpret the column as e.g. Int128, it
+                    /// may mean one of two things:
+                    ///  * The byte array is the 16 bytes of Int128, little-endian.
+                    ///  * The byte array is an ASCII string containing the Int128 formatted in base 10.
+                    /// There's no reliable way to distinguish these cases. We just guess: if the
```
Comment (cpcloud) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:767]: Forgive the drive-by comment, but don't Parquet logical types provide the ability to make this distinction?

Comment (al13n321) [reply src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:767]: No, they don't support integers wider than 64 bits. From `parquet.thrift`:
```
/**
 * Integer logical type annotation
 *
 * bitWidth must be 8, 16, 32, or 64.
 *
 * Allowed for physical types: INT32, INT64
 */
struct IntType {
  1: required i8 bitWidth
  2: required bool isSigned
}

[...]

/**
 * Decimal logical type annotation
 *
 * To maintain forward-compatibility in v1, implementations using this logical
 * type must also set scale and precision on the annotated SchemaElement.
 *
 * Allowed for physical types: INT32, INT64, FIXED, and BINARY
 */
struct DecimalType {
  1: required i32 scale
  2: required i32 precision
}

[...]

union LogicalType {
  1:  StringType STRING       // use ConvertedType UTF8
  2:  MapType MAP             // use ConvertedType MAP
  3:  ListType LIST           // use ConvertedType LIST
  4:  EnumType ENUM           // use ConvertedType ENUM
  5:  DecimalType DECIMAL     // use ConvertedType DECIMAL + SchemaElement.{scale, precision}
  6:  DateType DATE           // use ConvertedType DATE

  // use ConvertedType TIME_MICROS for TIME(isAdjustedToUTC = *, unit = MICROS)
  // use ConvertedType TIME_MILLIS for TIME(isAdjustedToUTC = *, unit = MILLIS)
  7:  TimeType TIME

  // use ConvertedType TIMESTAMP_MICROS for TIMESTAMP(isAdjustedToUTC = *, unit = MICROS)
  // use ConvertedType TIMESTAMP_MILLIS for TIMESTAMP(isAdjustedToUTC = *, unit = MILLIS)
  8:  TimestampType TIMESTAMP

  // 9: reserved for INTERVAL
  10: IntType INTEGER         // use ConvertedType INT_* or UINT_*
  11: NullType UNKNOWN        // no compatible ConvertedType
  12: JsonType JSON           // use ConvertedType JSON
  13: BsonType BSON           // use ConvertedType BSON
  14: UUIDType UUID           // no compatible ConvertedType
}
```

(Maybe we could write them as `DecimalType` with `scale = 0`, Decimal supports arbitrary length. The `precision` parameter would be awkward. E.g. UInt256 can hold some but not all 78-digit numbers. We could use `precision = 77` but allow the value to be greater than 10^77, but then other parquet decoders may consider this an error. We could use `precision = 78` and 33-byte-long value, but then our decoder can't know in advance whether all values will fit in UInt256. The latter seems ok though, maybe that would be the best way to do this. Also, signedness would probably be awkward in some way too.)

**Pull request merged**